File: archive/r/rust/josephus-problem.rs
use std::env::args;
use std::process::exit;
use std::str::FromStr;

fn usage() -> ! {
    println!("Usage: please input the total number of people and number of people to skip.");
    exit(0);
}

fn parse_int<T: FromStr>(s: &str) -> Result<T, <T as FromStr>::Err> {
    s.trim().parse::<T>()
}

// Reference: https://en.wikipedia.org/wiki/Josephus_problem#The_general_case
//
// Use zero-based index algorithm:
//
//    g(1, k) = 0
//    g(m, k) = [g(m - 1, k) + k] MOD m, for m = 2, 3, ..., n
//
// Final answer is g(n, k) + 1 to get back to one-based index
fn josephus(n: i32, k: i32) -> i32 {
    let mut g: i32 = 0;
    for m in 2..=n {
        g = (g + k) % m;
    }

    g + 1
}

fn main() {
    let mut args = args().skip(1);

    // Exit if 1st command-line argument not an integer
    let mut n: i32 = args
        .next()
        .and_then(|s| parse_int(&s).ok())
        .unwrap_or_else(|| usage());

    // Exit if 2nd command-line argument not an integer
    let mut k: i32 = args
        .next()
        .and_then(|s| parse_int(&s).ok())
        .unwrap_or_else(|| usage());

    // Run Josephus problem and store results
    println!("{}", josephus(n, k));
}

File: archive/c/c/josephus-problem.c
#include <stdio.h>
#include <stdlib.h>

int josephus(int n, int k) {
    if (n == 1)
        return 1;
    else
        return (josephus(n - 1, k) + k - 1) % n + 1;
}

int main(int argc, char *argv[]) {
    if (argc != 3) {
        printf("Usage: please input the total number of people and number of people to skip.\n");
        return 1;
    }

    char *endptr;
    int n = strtol(argv[1], &endptr, 10);
    if (*endptr != '\0' || n <= 0) {
        printf("Usage: please input the total number of people and number of people to skip.\n");
        return 1;
    }

    int k = strtol(argv[2], &endptr, 10);
    if (*endptr != '\0' || k <= 0) {
        printf("Usage: please input the total number of people and number of people to skip.\n");
        return 1;
    }

    int result = josephus(n, k);
    printf("%d\n", result);

    return 0;
}
